app/services/school_data_importer.rb

```diff
@@ -46,77 +41,119 @@ def school_data_files
     @gias_files
   end
 
-  def row_to_school(row)
-    school = School.find_or_initialize_by(urn: row.fetch("URN"))
-    school.name = row.fetch("EstablishmentName")
-    school.school_type_code = row.fetch("TypeOfEstablishment (code)").to_i
-    school.school_type_name = row.fetch("TypeOfEstablishment (name)")
-    school.address_line1 = row.fetch("Street")
-    school.address_line2 = row.fetch("Locality")
-    school.address_line3 = row.fetch("Town")
-    school.postcode = row.fetch("Postcode")
-    school.ukprn = row.fetch("UKPRN")
-    school.school_phase_type = row.fetch("PhaseOfEducation (code)").to_i
-    school.school_phase_name = row.fetch("PhaseOfEducation (name)")
-    school.school_website = row.fetch("SchoolWebsite")
-    school.school_status_code = row.fetch("EstablishmentStatus (code)").to_i
-    school.school_status_name = row.fetch("EstablishmentStatus (name)")
-    school.administrative_district_code = row.fetch("DistrictAdministrative (code)")
-    school.administrative_district_name = row.fetch("DistrictAdministrative (name)")
-    school.primary_contact_email = row.fetch("MainEmail")
-    school.secondary_contact_email = row.fetch("AlternativeEmail")
-    school.save!
-    school
-  end
+  class SchoolHandler
+    attr_reader :school, :data, :logger, :start_year
+
+    def initialize(school:, data:, logger:, start_year:, la_cache:, lad_cache:)
+      @school = school
+      @new_school = school.new_record?
+      @logger = logger
+      @data = data
+      @start_year = start_year
+      @la_cache = la_cache
+      @lad_cache = lad_cache
+    end
 
-  def row_to_local_authority(row)
-    local_authority = LocalAuthority.find_or_initialize_by(code: row.fetch("LA (code)"))
-    row_local_authority_name = row.fetch("LA (name)")
+    def call
+      ActiveRecord::Base.transaction do
+        record_school_details
 
-    if local_authority.persisted? && local_authority.name != row_local_authority_name
-      logger.info "LA name change in school import. Old name: #{local_authority.name}, New name: #{row_local_authority_name}"
+        if new_school?
+          link_school_to_local_authority
+          link_school_to_local_district_authority
+        end
+      end
     end
 
-    local_authority.name = row_local_authority_name
-    local_authority.save!
-    local_authority
-  end
+  private
 
-  def row_to_lad(row)
-    local_authority_district = LocalAuthorityDistrict.find_or_initialize_by(code: row.fetch("DistrictAdministrative (code)"))
-    row_lad_name = row.fetch("DistrictAdministrative (name)")
+    def new_school?
+      @new_school
+    end
+
+    def record_school_details
+      school.name = data.fetch("EstablishmentName")
+      school.address_line1 = data.fetch("Street")
+      school.address_line2 = data.fetch("Locality")
+      school.address_line3 = data.fetch("Town")
+      school.postcode = data.fetch("Postcode")
+      school.ukprn = data.fetch("UKPRN")
```

Comment:
I'd exclude UKPRN as well if they're not new, just to be safe.

Reply:
Done. :)